helix-core/src/main/java/org/apache/helix/model/ResourceConfig.java

```diff
@@ -389,6 +390,12 @@ public void setPreferenceLists(Map<String, List<String>> instanceLists) {
    * @throws IOException - when JSON conversion fails
    */
   public Map<String, Map<String, Integer>> getPartitionCapacityMap() throws IOException {
+    // It is very expensive to deserialize the partition capacity map every time this is called.
+    // Cache the deserialized map to avoid the overhead.
+    if (_deserializedPartitionCapacityMap != null) {
```

Comment:
we are relying on 'setPartitionCapacityMap' to reset the cache. Hopefully we are not calling setPartitinCapacityMap() too frequently.

Reply:
We do not call `setPartitionCapacityMap` anywhere but tests. Cache is refreshed by replacing ResourceConfig HelixProperty in dataProvider when there is a keyChange. Replacing the property will ensure that _deserializedPartitionCapacityMap will be recomputed if ResourceConfig partition capacity maps are changed.